### PR TITLE
topology updater: add e2e tests

### PR DESCRIPTION
**Pull request by fromanirh**
Builds on top of https://github.com/kubernetes-sigs/node-feature-discovery/pull/525 and add e2e tests for the new component.
Some refactoring was performed along the way to make room for the new tests; no intended changes in the behaviour of the existing tests.

Design Document: [here](https://docs.google.com/document/d/1Q-4wSu1tzmbOXyGk_2r5_mK6JdXXJA-bOd3cAtBFnwo/edit?ts=5f24171f#)
Issue: https://github.com/kubernetes-sigs/node-feature-discovery/issues/333
Documentation: https://github.com/kubernetes-sigs/node-feature-discovery/pull/526

### CODE REVIEW

**Comment by fromanirh** [comment]
reviewers please note the first 4 commits (up to `Expose Topology manager scope`, included) are part of https://github.com/kubernetes-sigs/node-feature-discovery/pull/525 , and are included here for PR consistency sake and to make CI pass.

**Comment by marquiz** [comment]
Rebase and drop the `vendor` directory

/ok-to-test

**Review by marquiz — CHANGES_REQUESTED** [review]
Drop the vendor/ dir

**Review by marquiz — CHANGES_REQUESTED** [review]
Thanks @fromanirh for the PR. I had some comments but I think this looks pretty straightforward.

One twist, though. In our test cluster kubelet config lives at `/etc/kubernetes/kubelet/kubelet-config.json` so we need to make the path configurable before we can merge the PR. I was thinking that maybe we should extend the e2e-test config file to accommodate this because it's basically a cluster-specific option. E.g. add a new top level `topologyUpdater` field under which to add this 🧐  WDYT? Even image pull policy could be added there (instead of Makefile) under `general` or smth, but I don't have strong opinions on that

**Comment by fromanirh** [comment]
Hey! just a quick update: being swamped with other things lately, but I will totally keep pushing this forward. On the upside, the code quality of the tests improved quite a lot and it should be reflected in the next respin of this PR, which will happen (as) soon (as possible). Sorry!

**Comment by marquiz** [comment]
Np, we're all swamped 😸  Thanks @fromanirh for the update and for working on this

**Comment by ArangoGutierrez** [comment]
@fromanirh any updates here? we would love to have better test coverage for next release cycle of NFD

**Comment by fromanirh** [comment]
@ArangoGutierrez yeah, this is totally on my TODO (and I agree about having better coverage) but had little to none capacity to work in this area. I'll try harder to reprioritize this or next week and I'll let you folks know.

File: test/e2e/node_feature_discovery.go
Comment (fromanirh) [reply test/e2e/node_feature_discovery.go:114]: fixed, thanks

**Review by fromanirh — COMMENTED** [review]
sorry for the (very very) late updates. But now I have some time allowance to move this forward.

**Comment by fromanirh** [comment]
/hold
need to read the CRD manifest as requested in the review

**Comment by fromanirh** [comment]
/hold cancel
addressed all review comments

**Comment by marquiz** [comment]
Thanks @fromanirh for the update! Great to have this (just reminds me about missing e2e tests in my own backlog 🙄 ). I took a quick look and no comments based on that. Now I'm just trying to get the tests pass in our e2e cluster 🤪 After that's resolved I'll take a final look and let's get this merged

**Comment by fromanirh** [comment]
thanks @marquiz . There most likely a bug I need to address. I'll work more on that and keep you posted. I think we're not far from the ready state.

File: test/e2e/utils/pod.go
Comment (marquiz) [inline test/e2e/utils/pod.go:326]: One more failure here: you need to create separate clusterrole and clusterrolebinding for the topologyupdater, matching what we have here ` deployment/base/rbac-topologyupdater/topologyupdater-clusterrole.yaml`.

**Comment by fromanirh** [comment]
> I did some debugging today. When is the last time you tried running these tests as there were typos etc that made it impossible for them to pass? wink I had a few comments but they are not enough, need more work...
> 
> Also, needs a rebase. Prolly better to wait for #839 before rebasing blush

Indeed, sorry about that: I should have marked WIP earlier. I need to do a deeper review to fix and update.
Again, sorry it take time of yours to run through the debug, I'll lift the WIP when I'm confident the tests are really up to date.

**Comment by marquiz** [comment]
@fromanirh any update on this one?

**Comment by fromanirh** [comment]
sure, going on work on this during this week

**Comment by fromanirh** [comment]
@marquiz *if* you agree with the refactoring of the e2e test utilities included in this PR, I split only that part in https://github.com/kubernetes-sigs/node-feature-discovery/pull/854. I think that if we like this direction, merging that part first makes sense and will also make the finalization of this PR easier. WDYT?

**Comment by marquiz** [comment]
> @marquiz _if_ you agree with the refactoring of the e2e test utilities included in this PR, I split only that part in #854. I think that if we like this direction, merging that part first makes sense and will also make the finalization of this PR easier. WDYT?

Yes, makes sense to me as well 👍

I think I had some comments in the utils part. Do you think you can address those in the separate PR?

**Comment by fromanirh** [comment]
> > @marquiz _if_ you agree with the refactoring of the e2e test utilities included in this PR, I split only that part in #854. I think that if we like this direction, merging that part first makes sense and will also make the finalization of this PR easier. WDYT?
> 
> Yes, makes sense to me as well +1
> 
> I think I had some comments in the utils part. Do you think you can address those in the separate PR?

absolutely!

File: test/e2e/utils/config.go
Comment (fromanirh) [reply test/e2e/utils/config.go:31]: I think I get what you mean, but we will require a helper getter because `test/e2e/node_feature_discovery.go` wants to access this data. I'll try something.

File: test/e2e/utils/config.go
Comment (fromanirh) [reply test/e2e/utils/config.go:26]: thanks, fixed.

File: test/e2e/utils/config.go
Comment (fromanirh) [reply test/e2e/utils/config.go:80]: thanks, fixed

File: test/e2e/utils/config.go
Comment (fromanirh) [reply test/e2e/utils/config.go:71]: I like this much more, going for it

File: test/e2e/utils/pod.go
Comment (fromanirh) [reply test/e2e/utils/pod.go:321]: thanks, fixed

File: test/e2e/utils/pod.go
Comment (fromanirh) [reply test/e2e/utils/pod.go:333]: it's probably the better place indeed. Going for it.

File: test/e2e/utils/pod.go
Comment (fromanirh) [reply test/e2e/utils/pod.go:342]: sure thing

File: test/e2e/utils/pod.go
Comment (fromanirh) [reply test/e2e/utils/pod.go:326]: done

**Comment by fromanirh** [comment]
Implemented all the changes. Still WIP because I need to verify locally. Not ready for review yet.

**Comment by fromanirh** [comment]
the PR is now reivewable again, but depends on https://github.com/kubernetes-sigs/node-feature-discovery/pull/854

**Review by marquiz — COMMENTED** [review]
Good job @fromanirh, I was now actually able to run the e2e-tests successfully in our test cluster 🥳

I left just a few comments but after those I think we're good to go with this

**Comment by marquiz** [comment]
We actually need to update our e2e-test-config in k8s test-infra before merging this (I'm on it)
/hold

File: test/e2e/utils/config.go
Comment (fromanirh) [reply test/e2e/utils/config.go:42]: sure thing, done

File: test/e2e/utils/pod.go
Comment (fromanirh) [reply test/e2e/utils/pod.go:46]: added, thanks for pointing this out.

File: test/e2e/utils/noderesourcetopology.go
Comment (fromanirh) [reply test/e2e/utils/noderesourcetopology.go:48]: added, thanks for pointing this out

File: test/e2e/topology_updater.go
Comment (fromanirh) [reply test/e2e/topology_updater.go:73]: Good point. Reworked a bit, please let me know what you think about it.

File: test/e2e/utils/pod.go
Comment (marquiz) [inline test/e2e/utils/pod.go:94]: nit 😄 

```suggestion
// DeletePodsAsync concurrently deletes all the pods in the given name:pod_object mapping. Returns when the longer operation ends.
```

File: test/e2e/utils/pod.go
Comment (marquiz) [inline test/e2e/utils/pod.go:43]: Hmm, this didn't catch my eye until now, but I realized that the image we use probably does not matter at all (just sleeping and waiting in our test cases). Would it be simpler to just use `k8s.gcr.io/pause`. Would simplify podSpec (no need to specify `Command:` and also virtually eliminate the need to re-pull (plus the image size when pulling is a lot smaller). WDYT?

File: test/e2e/utils/config.go
Comment (marquiz) [inline test/e2e/utils/config.go:55]: nit: missing "docstring" 🙄 Just caught my eye when skimming through. I know we are missing these in many places of the existing code but trying to improve things slowly...

**Review by marquiz — COMMENTED** [review]
Getting there. Just a few more notes.

**Comment by marquiz** [comment]
We also need https://github.com/kubernetes/test-infra/pull/27106 to get merged before merging this one.

File: test/e2e/utils/pod.go
Comment (fromanirh) [reply test/e2e/utils/pod.go]: thanks, fixed!

File: test/e2e/utils/pod.go
Comment (fromanirh) [reply test/e2e/utils/pod.go]: sure thing! changed as suggested.

File: test/e2e/utils/config.go
Comment (fromanirh) [reply test/e2e/utils/config.go:56]: no problem, fixed!

**Review by marquiz — APPROVED** [review]
Great persistence @fromanirh on this one 😊 I think this is good to go now 🎉
/lgtm

**Comment by marquiz** [comment]
...just need to get that test-infra pr merged in order to unhold this

**Comment by fromanirh** [comment]
Thanks for the reviews @marquiz !

**Pull request merged**
